### ISSUE TITLE
Graphics problems in Chrome & Electron apps in macOS 12 guest on M1 Mac

### ISSUE THREAD

**Issue by user425846** [utmapp/UTM#3692]
**Describe the issue**
Generally the performance inside the VM is really really good, but some apps just dont work correctly. A few examples:

<img width="936" alt="Bildschirmfoto 2022-02-27 um 03 08 12" src="https://user-images.githubusercontent.com/100499696/155865664-b52c9141-28bc-4c34-b0ea-d465bbba3ef7.png">

Permanent graphical Glitches in Visual Studio Code. Also observed in Exodus Wallet, Monero Gui Wallet and Sublime Text. App is unresponsive during these glitches and that doesnt change.

<img width="936" alt="Bildschirmfoto 2022-02-27 um 03 18 20" src="https://user-images.githubusercontent.com/100499696/155865680-c466a19a-a467-4a9b-bc9a-b4e4fbe99c59.png">

Rider IDE for example is showing a crash (Probably something on their end). But if you continue, same issue as above, its glitchy and unresponsive.

Visual Studio 22 Preview doesnt open at all, Visual Studio 2019 runs perfectly fine and the performance is great.

**Configuration**
* UTM Version: 3.1.1 (48)
* OS Version: 12.2.1
* Intel or Apple Silicon? Apple Silicon

**Debug log**
I think there is no debug option available for this guest? At least i didnt find one, sorry!

**Comment by LeonSolisPedro** [comment]
I have this problem too! Its probably on apple's side. I couldn't find anything on the macos release notes page, and neither in their forums

Its so annoying!

**Comment by kjunggithub** [comment]
I get the same thing with Chrome. Maybe electron based apps just don't work at all.

**Comment by alxsoft** [comment]
You can launch these apps with disabled gpu acceleration from Terminal:
open /Applications/Visual\ Studio\ Code.app --args --disable-gpu

**Comment by user425846** [comment]
> You can launch these apps with disabled gpu acceleration from Terminal: open /Applications/Visual\ Studio\ Code.app --args --disable-gpu

This definitely fixed it! I didnt know you could do that. Obviously its still not a great experience, its not running super smoothly without the 3d acceleration it usually gets, but at least its usable now!

**Comment by user425846** [comment]
> I get the same thing with Chrome. Maybe electron based apps just don't work at all.

Yeah thats what i figured aswell. Most problems i had were with electrum apps. As chrome itself also has the same issue, i guess the issue here is with chromium.

But i also get a milder form of these issues with safari. The only browser that ive tested which works perfectly so far is firefox.

**Comment by kjunggithub** [comment]
> > I get the same thing with Chrome. Maybe electron based apps just don't work at all.
> 
> Yeah thats what i noticed aswell. Most problems i had were with electron apps. As chrome itself also has the same issue, i guess the issue here is with the chromium render engine.
> 
> But i also get a milder form of these issues with safari. The only browser that ive tested which works perfectly so far is firefox.

I get the same thing. Firefox based browsers all work good. (tested Tor Browser and LibreWolf)

**Comment by kababoom** [comment]
I think it's got something to do with the video driver..

**Comment by LeonSolisPedro** [comment]
I began to question whether this is a problem on Apple's end or not (I did not find any GitHub issues on the Electron or VSCode Repository) I filed a feedback to apple anyways, they seems to be aware of the issue somehow, but I haven't received a reply back.


About visual Studio..., I managed to run Visual Studio for Mac 2022 by running this commands

```
export COMPlus_ReadyToRun=0
open /Applications/Visual\ Studio\ \(Preview\).app
```

This is just a [workaround](https://github.com/dotnet/runtime/issues/64103#issuecomment-1079259263), Its probably going to be fixed in future releases

**Comment by LeonSolisPedro** [comment]
I started to wonder if the Electron problem is on the Apple's end or not.
I filed a feedback to apple anyways, they seems to be aware of the issue somehow, but I haven't received a reply back.

About Visual Studio... I managed to find a workaround to run Visual Studio 2022 on an Emulated M1 Mac, you have to run these commands on the Terminal:

```
export COMPlus_ReadyToRun=0
open /Applications/Visual\ Studio\ \(Preview\).app
```

However, I am unaware of any drawbacks by doing this.
A [GitHub contributor explained](https://github.com/dotnet/runtime/issues/64103#issuecomment-1079259263) very detailed about why Dotnet 6 was crashing on M1 Virtual Machines.

**Comment by handletastic** [comment]
I can confirm that disabling the gpu works as a workaround.

**Comment by jukrb0x** [comment]
upvote for bugfix.

**Comment by zelnaut** [comment]
I haven't had any issue with VSCode using virtio-ramfb-gl (GPU) display card, although Chrome isn't rendering correctly for me.

Seems this is the relevant section of the Chrome logs.

```
[3315:3315:0712/014846.061113:WARNING:angle_platform_impl.cc(48)] ShaderGL.cpp:206 (checkShader): 
0:1(12): warning: extension `GL_ARB_gpu_shader5' unsupported in fragment shader
0:2(12): warning: extension `GL_EXT_gpu_shader5' unsupported in fragment shader
0:3(1): error: `out' qualifier in declaration of `webgl_SecondaryFragColor' only valid for function parameters in GLSL 1.10
0:0(0): error: fragment shader writes to both `gl_FragColor' and `webgl_SecondaryFragColor'

[3315:3315:0712/014846.065810:ERROR:shared_context_state.cc(76)] Skia shader compilation error
------------------------
#version 100

#extension GL_EXT_blend_func_extended : require
precision mediump float;
precision mediump sampler2D;
uniform sampler2D uTextureSampler_0_S0;
varying highp vec2 vTextureCoords_S0;
varying highp float vTexIndex_S0;
varying mediump vec4 vinColor_S0;
void main() {
    mediump vec4 outputColor_S0;
    outputColor_S0 = vinColor_S0;
    mediump vec4 texColor;
    {
        texColor = texture2D(uTextureSampler_0_S0, vTextureCoords_S0, -0.5);
    }
    mediump vec4 outputCoverage_S0 = texColor;
    {
        outputCoverage_S0.w = max(max(outputCoverage_S0.x, outputCoverage_S0.y), outputCoverage_S0.z);
        gl_SecondaryFragColorEXT = outputColor_S0.w * outputCoverage_S0;
        gl_FragColor = outputColor_S0 * outputCoverage_S0;
    }
}
```

**Comment by wakco** [comment]
Another upvote for bug fix, as the workaround stopped working, at least from Google Chrome at around version 100 of Google Chrome. (version 105 at time of posting).

**Comment by julbd** [comment]
If no proper fix is found, could you allow to disable gpu acceleration on system level ? The Standard Notes app, for example, looks like an Electron app but does not respond to the `--args --disable-gpu` hack.

**Comment by ElevateAir** [comment]
Same issue here, on M1 MacOS on MacOS, VS Code does not work even with disabled GPU acceleration both from the command line and from the application itself. Really sad as it's one of my core applications. Interestingly Chrome works very well and so far VS Code was the only app that had issues for me.

PS: As an alternative to VS Code Atom seems to work pretty well.

**Comment by jukrb0x** [comment]
I found it interesting that the same issue goes to Steam for macOS, but the problem only occurs after switching between other apps and Steam and does not always arise, guess they originate from the same graphic issue.

**Comment by LeonSolisPedro** [comment]
Visual Studio 17.4 is now fully functional on a Virtual M1 Mac

**Comment by acherkashin** [comment]
I had issues with both Firefox and VS Code.

I used [the following tutorial](https://www.youtube.com/watch?v=1WWj6qoWhJw&ab_channel=KskRoyal ) to install Ubuntu on Mac with M1. Image was downloaded from [here](https://cdimage.ubuntu.com/jammy/daily-live/current/).

To fix the issue with Firefox and VS Code I've used the [following article](https://linuxconfig.org/how-to-disable-wayland-and-enable-xorg-display-server-on-ubuntu-18-04-bionic-beaver-linux).  To be short - I've disabled `Wayland` in the following way:

1. Type in terminal - `sudo nano /etc/gdm3/custom.conf`
2. Uncomment `WaylandEnable=false` by removing `#` in front of it
3. If you have sounds problem, just make sure you've turned it on on Linux

<img width="340" alt="image" src="https://user-images.githubusercontent.com/9947582/205232708-b1205b50-bad5-44dc-b18c-e1c052730f31.png">

**Comment by acherkashin** [comment]
But anyway, vscode looks in the following way. Icons and some texts are not shown.

<img width="1438" alt="image" src="https://user-images.githubusercontent.com/9947582/205497921-9da25b08-2743-41b1-8d61-ec5828670a32.png">

At the same VS Coding works fine

<img width="1435" alt="image" src="https://user-images.githubusercontent.com/9947582/205498114-fafea8ef-429c-425e-a9fc-1217cd298466.png">

**Comment by ElevateAir** [comment]
Unfortunately I found the same issues apply to the google drive app :(  login screen does not display in macOS.

**Comment by mahnunchik** [comment]
I've face with graphic issue on macOS host and Windows 11 (22000.1098) arm64 guest while running Steam app.
